### DISCUSSION TITLE
Efficient Round Function

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#208] [Show and tell]
There is no math.round function.

I found this nice little round function that only uses addition

```
function round(num)
    return num + (2^52 + 2^51) - (2^52 + 2^51)
end
```

**Comment by dnaldoog** [comment]
That's interesting I don't really understand it :(

I pulled this one off a forum somewhere, years ago:

<pre><code>
function round(num, numDecimalPlaces) -- rounding function
  local mult = 10^(numDecimalPlaces or 0)
return math.floor(num * mult + 0.5) / mult
end
</pre></code>

but I got to thinking and came up with this: Does it work?

<pre><code>
function round(num)
return=num+0.5 >= math.ceil(num) and math.ceil(num) or math.floor(num)
end
</pre></code>

**Comment by Godlike-Productions** [comment]
This is the explanation from https://stackoverflow.com/questions/18313171/lua-rounding-numbers-and-then-truncate

It takes advantage of the limits of double precision numbers and FPU rounding.

Adding and subtracting 2^52 to any number between 0 and 2^52-1 rounds it by pushing it to the range 2^52..2^53, where the double precision floats lose all their decimals, before adjusting back. To cover negatives, I added 2^51 in order to ensure the intermediate result is still in the range 2^52..2^53.

**Comment by dnaldoog** [comment]
Or on that page the even more succinct 

<pre><code>
function round(num)
return math.floor(num+0.5)
end
</pre>
</code>

and another one on that page similar to the one I came up with using lua's ternary operatory emulation

<pre>
<code>
function round(x)
  return x>=0 and math.floor(x+0.5) or math.ceil(x-0.5)
end
</pre>
</code>

**Comment by Godlike-Productions** [comment]
I do 't think they work in all cases. 0.49 for example returns 1.

**Reply by dnaldoog** [reply]
I can't duplicate that problem (?)

[God Like Round Function_1_0_0_2021-03-09_20-29.zip](https://github.com/RomanKubiak/ctrlr/files/6108130/God.Like.Round.Function_1_0_0_2021-03-09_20-29.zip)

![aganza](https://user-images.githubusercontent.com/76831683/110476123-fc803b00-8124-11eb-9e50-29743a7ae71a.png)
